src/py21cmfast/src/SpinTemperatureBox.c

```diff
@@ -1,2760 +1,1932 @@
-                        LOG_ERROR("One of the frequency interpolation tables has an infinity or a NaN");
-                        Throw(TableGenerationError);
+                    xray_sfr = (sfr_term*astro_params->L_X + sfr_term_mini*astro_params->L_X_MINI);
+                    xidx = m_xHII_low_box[box_ct];
+                    ival = inverse_val_box[box_ct];
+                    dxheat_dt_box[box_ct] += xray_sfr * xray_R_factor * (freq_int_heat_tbl_diff[xidx][R_ct] * ival + freq_int_heat_tbl[xidx][R_ct]);
+                    dxion_source_dt_box[box_ct] += xray_sfr * xray_R_factor * (freq_int_ion_tbl_diff[xidx][R_ct] * ival + freq_int_ion_tbl[xidx][R_ct]);
+                    dxlya_dt_box[box_ct] += xray_sfr * xray_R_factor * (freq_int_lya_tbl_diff[xidx][R_ct] * ival + freq_int_lya_tbl[xidx][R_ct]);
+                    dstarlya_dt_box[box_ct] += sfr_term*dstarlya_dt_prefactor[R_ct] + sfr_term_mini*starlya_factor_mini; //the MINI factors might not be allocated
+
+                    if(flag_options->USE_LYA_HEATING){
+                        dstarlya_cont_dt_box[box_ct] += sfr_term*dstarlya_cont_dt_prefactor[R_ct] + sfr_term_mini*lyacont_factor_mini;
+                        dstarlya_inj_dt_box[box_ct] += sfr_term*dstarlya_inj_dt_prefactor[R_ct] + sfr_term_mini*lyainj_factor_mini;
                     }
                 }
             }
         }
+    }
+    //R==0 part
+#pragma omp parallel private(box_ct)
+    {
+        double curr_delta;
+        int print_count=0;
+        struct Ts_cell ts_cell;
+        struct Box_rad_terms rad;
+        #pragma omp for reduction(+:J_alpha_ave,xheat_ave,xion_ave,Ts_ave,Tk_ave,x_e_ave,eps_lya_cont_ave,eps_lya_inj_ave)
+        for(box_ct=0; box_ct<HII_TOT_NUM_PIXELS; box_ct++){
+            curr_delta = perturbed_field->density[box_ct] * growth_factor_zp * inverse_growth_factor_z; //map density to z' TODO: look at why this is an option
+            //this corrected for aliasing before, but sometimes there are still some delta==-1 cells
+            //  which breaks the adiabatic part, TODO: check out the perturbed field calculations to find out why
+            if (curr_delta <= -1){
+                curr_delta = -1+FRACT_FLOAT_ERR;
+            }
+
+            //NOTE: if the filtered grids have sfr density, and the factors convert to emissivity density, surely I need a 1/(1+delta_curr)
+            //  factor here for absorber density (i.e overdense regions need more radiation to have the same effects)
+            //  What am I missing? The ionisation box has a final delta dependence of (1+delta_source)/(1+delta_absorber) which makes sense
+            //  But here it's just (1+delta_source).
+            //  mean_sfr_zpp has no rhoc (divided out), The tables have no rhoc, there is only one in the constant factor, and a 1+delta_source in the grids
```

Comment:
maybe another question for @qyx268

Reply:
We figured this one out when I was in Melbourne. Since we want to avoid the 6D calculation of calculating the attenuation by each cell towards each cell, we assume that all attenuation happens at mean density. This means that we _must_ also assume that all absorption happens at mean density, otherwise photons will be gained/lost throughout the calculation.